doc/manuals/sprokit/embedded_pipeline.rst

```diff
@@ -0,0 +1,327 @@
+
+    // Data values need to be supplied to all connected ports
+    // (based on the previous pipeline definition)
+    ds.add_value( "port2", i );
+    ds.add_value( "port3", i );
+    ep.send( ds ); // push into pipeline
+
+    // Get output from pipeline
+    auto rds = ep.receive();
+
+    // get value from the output adapter
+    int val = get_port_data<int>( "out_num" );
+  } // end for
+
+  ep.send_end_of_input(); // indicate end of input
+
+  auto rds = ep.receive(); // Retrieve end of input data item.
+  if ( ! ep.at_end() || ! rds.is_end_of_data() )
+  {
+    // This is unexpected. Both conditions should be true.
+  }
+
+
+The following section builds a pipeline at runtime using a set of
+macros to simplify constructing the pipeline definition. The full list
+of these macros and associated documentation can be found in the
+documentation for the include file. ::
+
+    1)    #include <sprokit/pipeline_util/literal_pipeline.h>
+
+      // SPROKIT macros can be used to create pipeline description
+    2)    std::stringstream pipeline_desc;
+    3)    pipeline_desc << SPROKIT_PROCESS( "input_adapter",  "ia" )
+    4)                  << SPROKIT_PROCESS( "output_adapter", "oa" )
+
+    5)                  << SPROKIT_CONNECT( "ia", "counter",  "oa", "out_num" )
+    6)                  << SPROKIT_CONNECT( "ia", "port2",    "oa", "port3" )
+    7)                  << SPROKIT_CONNECT( "ia", "port3",    "oa", "port2" );
+
+Line 1 includes the file that defines a set of macros that can be used
+to pragmatically create a pipeline definition.
+
+Line 2 defines a string stream that will contain the constructed
+pipeline definition.
+
+Line 3 defines a process of type "input_adapter" that will be
+referenced as "ia" in the pipeline definition.
+
+Line 4 defines a process of type "output_adapter" that will be
+referenced as "oa" in the pipeline definition.
+
+Line 5 connects port "counter" on process "ia" (the input adapter) to
+port "out_num" on process "oa" (the output adapter).
+
+Lines 5 and 6 make additional connections between the input and output
+adapter.
+
+The folloing section created and starts the pipeline.::
+
+    // create embedded pipeline
+    8)  kwiver::embedded_pipeline ep;
+    9)  ep.build_pipeline( pipeline_desc );
+
+    // Start pipeline
+   10)  ep.start();
+
+Line 8 creates the embedded pipeline object.
+
+Line 9 builds the pipeline based on the supplied input stream. Errors
+may be detected while building the pipeline.
+
+Line 10 starts the pipeline running. Control returns after the pipeline
+is started to allow this thread to optionally supply inputs and/or
+consume outputs while the pipeline runs asynchronously.
+
+The following code illustrates how data items are supplied to the
+embedded pipeline. In this sample code, ten sets of data are sent to
+the pipeline with the result being read back immediately. This is not
+always that realistic for more complicated pipelines because there is
```

Comment:
I'm not sure "realistic" here is best conveying the desired meaning. A better phrasing might be "This may be impractical for more complicated pipelines, because some operations require more than one input before an output can be computed. It is also not very efficient, since sending many inputs will allow the pipeline to keep working while we wait on an output."